### PR TITLE
[quotemark] Excuse more backtick edge cases

### LINKED ISSUE #4645: [quotemark] backtick breaks old typescript
### Bug Report

- __TSLint version__: 5.15.0
- __TypeScript version__: 2.4.x
- __Running TSLint via__: Atom

#### TypeScript code being linted

```ts
declare let v: string | number

if (typeof v === "string") {
    v.substring(1)
}

// and

let obj: {
    key: "literal"
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "quotemark": [true, "backtick"]
    }
}
```

#### Actual behavior

Quotes are flagged (and changed to backticks with --fix)

#### Expected behavior

The quotes remain as they are, because this version of TS does not narrow the type of v when using `typeof` with a backtick string, and does not compiler at all when using a backtick string as a literal type

### PR DESCRIPTION
**Pull request by ericbf**
This edge cases were previously flagged when they should be ignored, as changing them breaks typescript. This commit makes it so they are ignored. It also organizes a little better, using functions instead of multi-layered conditionals (it was getting confusing).

#### PR checklist

- [x] Addresses an existing issue: #4588, fixes #4645
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:

I keep finding other edge cases that I previously missed. I *think* this is all of them now, but we shall see. Related to #4588.

#### Is there anything you'd like reviewers to focus on?

See if y'all find any other edge cases.

#### CHANGELOG.md entry:

[bugfix] Excuse more quotemark backtick edge cases

### CODE REVIEW

File: src/rules/noInferredEmptyObjectTypeRule.ts [hunk 1]
```diff
@@ -68,6 +68,7 @@ class NoInferredEmptyObjectTypeRule extends Lint.AbstractWalker {
             } else if (node.kind === ts.SyntaxKind.NewExpression) {
                 this.checkNewExpression(node as ts.NewExpression);
             }
+
```
Comment (JoshuaKGoldberg) [inline src/rules/noInferredEmptyObjectTypeRule.ts]: Unrelated changes; please revert.

File: src/rules/quotemarkRule.ts [hunk 2]
```diff
@@ -245,3 +246,91 @@ function getJSXQuotemarkPreference(
...
+        node.parent.kind !== ts.SyntaxKind.PropertyAssignment &&
+        node.parent.kind !== ts.SyntaxKind.MethodDeclaration
+    ) {
+        // If the node is neither a property assignment or method declaration, it's not a name in an assignment
+        return false;
+    }
+
+    return (
+        // In typescript below 2.7.1, don't change values either
+        ts.version < "2.7.1" ||
```
Comment (JoshuaKGoldberg) [inline src/rules/quotemarkRule.ts]: Directly checking `ts.version` is a little unusual... maybe there's a cleaner way? Thoughts @adidahiya? 

Also, @ericbf, could you elaborate on why this is necessary?

File: src/rules/noInferredEmptyObjectTypeRule.ts (hunk 1 above)
Comment (ericbf) [reply src/rules/noInferredEmptyObjectTypeRule.ts]: Gladly. Just making all merge checks pass.

File: src/rules/quotemarkRule.ts (hunk 2 above)
Comment (ericbf) [reply src/rules/quotemarkRule.ts]: Typescript below the listed version fails compilation (or otherwise misbehaves) when backticks are used in some places where it succeeds in future versions. For instance, ``typeof str === `string``` does not narrow the type of `str` in TS versions below 2.7.1, causing compilation errors. In the other cases, older TS throws errors where NoSubstitutionLiterals are used instead of single/double quoted strings, but works fine in above 2.7.1; rather than cripple the rule to support order versions, I split it up.

Comment (ericbf) [reply src/rules/quotemarkRule.ts]: See the `typeof` issue [here](https://github.com/Microsoft/TypeScript/issues/20949). Its milestone was 2.7 (hence the split in this PR).
And [here's](https://github.com/Microsoft/TypeScript/issues/16592) the issue for backticks strings as types. This one was 2.6, but rather than doing multiple splits, I just did one at 2.7.

Comment (ericbf) [reply src/rules/quotemarkRule.ts]: Created corresponding issue, #4645

Comment (JoshuaKGoldberg) [reply src/rules/quotemarkRule.ts]: Thanks! This LGTM but before merging I'd like to hear from adidahiya on whether there's a cleaner way to check on these things.

**Comment by ericbf** [comment]
Any update here? @adidahiya? @JoshuaKGoldberg?

**Comment by ericbf** [comment]
Any update here?

File: src/rules/quotemarkRule.ts (hunk 2 above)
Comment (adidahiya) [reply src/rules/quotemarkRule.ts]: I agree it's unusual and should be avoided if possible, but this seems like a reasonable use case for checking ts.version.

However, can you please use the semver package and the helper function from `parse.ts` to check version ranges instead?

```ts
import { lt } from "semver";
import { getNormalizedTypescriptVersion } from "../../verify/parse";

function hasOldTscBacktickBehavior() {
    return lt(getNormalizedTypescriptVersion(), "2.7.1");
}
```

**Comment by adidahiya** [comment]
Also if you merge master you'll get the new required CI check, "clean-lockfile"
